src/srcset.js

```diff
@@ -180,119 +165,110 @@ export class Srcset {
+    let minWidth = Infinity;
+
+    for (let i = 0; i < sources.length; i++) {
+      const sWidth = sources[i].width;
+      const score = Math.abs(sWidth - width);
+
+      // Select the one that is closer with a slight preference toward larger
+      // widths. If smaller size is closer, enforce minimum ratio to ensure
+      // image isn't too distorted.
+      if (score <= minScore * 1.1 || width / minWidth > 1.2) {
+        minIndex = i;
+        minScore = score;
+        minWidth = sWidth;
+      } else {
+        break;
       }
-      prevWidth = sourceWidth;
     }
-    // Use the first (maximum) value.
-    return 0;
+    return minIndex;
   }
 
   /**
-   * @param {number} _width
    * @param {number} dpr
    * @return {number}
    * @private
    */
-  selectByDpr_(_width, dpr) {
-    let minIndex = -1;
-    let minScore = 1000000;
-    for (let i = 0; i < this.sources_.length; i++) {
-      const source = this.sources_[i];
-      // Default DPR = 1.
-      const sourceDpr = source.dpr || 1;
-      const score = Math.abs(sourceDpr - dpr);
-      if (score < minScore) {
-        minScore = score;
+  selectByDpr_(dpr) {
+    const sources = this.sources_;
+    let minIndex = 0;
+    let minScore = Infinity;
+
+    for (let i = 0; i < sources.length; i++) {
+      const score = Math.abs(sources[i].dpr - dpr);
+      if (score <= minScore) {
         minIndex = i;
+        minScore = score;
+      } else {
+        break;
       }
     }
     return minIndex;
   }
 
   /**
-   * Returns the last source in the srcset, which is the default source.
-   * @return {!SrcsetSourceDef}
+   * Returns all URLs in the srcset.
+   * @return {!Array<string>}
    */
-  getLast() {
-    return this.sources_[this.sources_.length - 1];
-  }
-
-  /**
-   * Returns all sources in the srcset.
-   * @return {!Array<!SrcsetSourceDef>}
-   */
-  getSources() {
-    return this.sources_;
+  getUrls() {
+    return this.sources_.map(s => s.url);
   }
 
   /**
    * Reconstructs the string expression for this srcset.
+   * @param {function(string):string=} opt_mapper
    * @return {string}
    */
-  stringify() {
+  stringify(opt_mapper) {
     const res = [];
-    for (let i = 0; i < this.sources_.length; i++) {
-      const source = this.sources_[i];
-      if (source.width) {
-        res.push(`${source.url} ${source.width}w`);
-      } else if (source.dpr) {
-        res.push(`${source.url} ${source.dpr}x`);
+    const sources = this.sources_;
+    for (let i = 0; i < sources.length; i++) {
+      const source = sources[i];
+      let src = source.url;
+      if (opt_mapper) {
+        src = opt_mapper(src);
+      }
+      if (this.widthBased_) {
+        src += ` ${source.width}w`;
       } else {
-        res.push(`${source.url}`);
+        src += ` ${source.dpr}x`;
       }
+      res.push(src);
     }
     return res.join(', ');
   }
 }
 
 function sortByWidth(s1, s2) {
   user().assert(s1.width != s2.width, 'Duplicate width: %s', s1.width);
```

Comment:
Is this sorting change intentional? 0.0

Reply:
Yes, so we can stop backwards iterating.